scipy/stats/_stats.pyx

```diff
@@ -513,6 +513,94 @@ cdef double _geninvgauss_pdf(double x, void *user_data) nogil except *:
     return math.exp(_geninvgauss_logpdf_kernel(x, p, b))
 
 
+cdef double _phi(double z) nogil:
+    cdef double inv_sqrt_2pi = 0.3989422804014327
+    return inv_sqrt_2pi * math.exp(-0.5 * z * z)
+
+
+cdef double _Phi(double z) nogil:
+    cdef double m_sqrt1_2 = 0.7071067811865475
+    return 0.5 * math.erfc(-z * m_sqrt1_2)
+
+
+cdef double _genstudentized_range_cdf(int n, double[2] x, void *user_data) nogil:
+    # evaluates the integrand of Equation (3) by Batista, et al [2]
+    # destined to be used in a LowLevelCallable
+    q = (<double *>user_data)[0]
+    k = (<double *>user_data)[1]
+    df = (<double *>user_data)[2]
+
+    s = x[1]
+    z = x[0]
+
+    # terms are evaluated within logarithms when possible to avoid overflows
+    log_terms = math.log(k) + (df / 2) * math.log(df) \
+               - (math.lgamma(df / 2) + (df / 2 - 1) * math.log(2)) \
+               + (df - 1) * math.log(s) - (df * s * s / 2) \
+               + math.log(0.3989422804014327) - 0.5 * z * z  # phi estimation.
+
+    # this term is excluded from the logarithm as it results in zero division
```

Comment:
This warning is thrown during integration if that term is included in the log:

![err](https://user-images.githubusercontent.com/45153623/114341136-e8f23680-9b0d-11eb-97b9-fd09d8764076.PNG)
```
RuntimeWarning: divide by zero encountered in ?
```

Should we include the term in the log and just suppress the error? It doesn't seem like a huge deal b/c the number comes out exactly the same.

Reply:
I don't see divide but zero... but when `s` is zero, we have 0^(k-1) = 0, and `np.log(0)` is negative infinity. Can you check whether that's the problem?

Reply:
Interesting... After a good deal of trial and error, it seems like `z` is actually the main influencer of when the log of this term causes a warning. Limiting execution of it to a small range of `s` (IE only execute for `2 < s < 2.1`) still results in the error, while limiting `z` (I tested `-1 < z < 1`) removes it. It seems like there are multiple values per integral that result in this warning as well - both `z < 0` and `z > 0` cause it to show. 

Regardless, you're right that it happens when the term inside the log (`_Phi(z + q * s) - _Phi(z)`) is 0. I've confirmed that with this snippet:
```python
import math
import numpy as np
import scipy.integrate as integrate

def _phi(z):
    """estimates the normal PDF. Used in `studentized range`"""
    inv_sqrt_2pi = 0.3989422804014327
    return inv_sqrt_2pi * math.exp(-0.5 * z * z)


def _Phi(z):
    """estimates the normal CDF. Used in `studentized range`"""
    m_sqrt1_2 = 0.7071067811865475
    return 0.5 * math.erfc(-z * m_sqrt1_2)


def cdf(q, k, df):
	def _genstudentized_range_cdf(z, s):
	
		# harcoded constants to avoid useless log evals
		log_2 = 0.6931471805599453
		log_inv_sqrt_2pi = -0.9189385332046727
		
		print(q, k, df, s, z, _Phi(z + q * s) - _Phi(z))
		
		# terms are evaluated within logarithms when possible to avoid overflows
		log_terms = (math.log(k) + (df / 2) * math.log(df)
					 - (math.lgamma(df / 2) + (df / 2 - 1) * log_2)
					 + (df - 1) * math.log(s) - (df * s * s / 2)
					 + log_inv_sqrt_2pi - 0.5 * z * z)  # phi estimation.


		# this term is excluded from the logarithm as it results in zero division
		if True:
			return math.exp(log_terms + (k - 1) * math.log(_Phi(z + q * s) - _Phi(z)))
		else:
			return math.exp(log_terms)
			
	ranges = [(-np.inf, np.inf), (0, np.inf)]
	opts = dict(epsabs=1e-11, epsrel=1e-12)

	return integrate.nquad(_genstudentized_range_cdf, ranges=ranges, opts=opts)[0]
	
print(cdf(3, 3, 10))
```

The question now is whether to implement a check for this failure condition within the python function or whether to just ignore the warning.

Reply:
For now, I'll leave it (almost) as-is. The term can't get "big" because it is limited between 0-1 because `_Phi()` is the normal CDF, so placing it in a log probably doesn't give a huge benefit.

Reply:
As we discussed, let's calculate this term in a smarter way based on discussion in gh-13218

Reply:
I tried the solution mentioned in that PR, but it doesn't seem applicable to normal CDF evaluations. Simplifying it to remove the erf-specific scale factors leaves `np.exp(log_ndtr(z1))-np.exp(log_ndtr(z0))`, which can be further simplified to `ndtr(z1) - ndtr(z0)`, which is what was being evaluated before. I get the impression that precision will be limited no matter how this is evaluated because it ultimately needs to pass through an non-logarithmic state to do the subtraction. The warning still appeared using this as well.

Along different lines, I tried replacing any 0s that cropped up with a machine epsilon, which gave correct answers but was pretty messy implementation-wise. I also tried evaluating the difference on the side of the CDF closest to 0, but that also resulted in the warning.

Ultimately, I decided that suppressing the warning with `with np.errstate(divide='ignore'):` is probably the cleanest way to solve the issue, barring any suggestions to the contrary. It still results in the same answer, and it allows cleaner code in the Cython function.

Reply:
It turns out that `np.exp(log_ndtr(*))` lets you go a tiny but further out in the tail before becoming indistinguishable from zero.
```
> ndtr(-27*np.sqrt(2))
0
> np.exp(log_ndtr(-27*np.sqrt(2)))
2.61855e-319
``` 
The more important idea is that ndtr(*) loses precision quicker toward the right because the magnitude gets really close to 1 instead of really close to 0. 
```
> ndtr(9)  # indistinguishable from 1
1.0
> ndtr(-9)  # easily distinguishable from 1
1.1285884059538324e-19
```

The idea is to either use the _complimentary_ error function (1-erf) or use the fact that the calculation is symmetric about zero to move the calculation to the left tail.

It sounds like this is still not enough for this application, though.